text/0000-module-resolution.md

```diff
@@ -0,0 +1,458 @@
+        is installed and can be resolved (by using Node `node_modules` 
+        algorithm). Then `GoTo 1` recursively merging all the 
+        ModuleRecordEntries collecting the scope forEach as npm 
+        packages are traversed.
+3.2.4 Otheriwse throw an invalid ModuleRecord error
+    3.3 Return all the ModuleRecordEntries collected.
+```
+
+### Algorithm to resolve a module in iterable way
+
+This algorithm is very similar to the [NodeJS require](https://nodejs.org/api/modules.html#modules_all_together),
+the main difference is that we don't traverse upwards all module paths. 
+
+```
+Lets `importee` be module specifier to be resolved and `importer` the path of the module on 
+which is being imported from.
+
+1. Find the closest lwc.config.js or package.json that contains an "lwc" config by traversing 
+upwards in the file system.
+
+2. Match the current importee againts a ModuleRecord defined in the modules array.
+    2.1 If the ModuleRecord is not found throw since resolution
+
+3. For each ModuleRecord
+    3.1 Match the `importee` to a module specififer:
+        3.1.1 If is an `AliasModuleRecord` validate the path and add
+        match the ModuleRecordEntry with:
+          - `specifier` as the `importee` value.
+          - `entry` with the `importer` value.
+
+        3.1.2 Else if is a `DirModuleRecord` validate the path and find of 
+        the modules that match the structure:
+        [namespace]/[componentName]/[componentName.{html|css|js|ts}] and match a 
+        ModuleRecordEntry with:
+          - `specifier` be `[namespace]/[componentName]`
+          - `entry` be `[dir]/[namespace]/[componentName]/[componentName.{html|css|js|ts}]`
+
+        3.1.3 Else if is a `NpmModuleRecord`, find lwc.config.js or package.json lwc config
+            3.1.3.1 Check that the component is exposed.
+```
+
+#### Future ModuleRecord types
+
+We will not discard adding another types of module resolutions, like for example, find modules directly from git or even from a url.
+This schema allows us to add new types as we see fit.
+
+## Scoping
+
+Its an invariant of the module resolution system to allow for multiple versions of multiple components within a particular application configuration. As the modules are resolved, we keep track of the context on which those modules appear, basically who is the importer (and its path). You can check again the [basic example](#basic_example) on the top.
+
+This concept of scope matches the [current specification of import maps](https://github.com/WICG/import-maps#multiple-versions-of-the-same-module) which will allow us to create the equivalent resolution for the browser.
+
+## Resolving modules at build time.
+
+This algorithm can be used and integrated with tools like Webpack or rollup, in fact we have created both the `@lwc/module-resolver` and the `@lwc/rollup-plugin` packages that we officially support. Other projects in the ecosystem such as `create-lwc-app` also leverage those primitives.
```

Comment:
I think it might be interesting to note that if the LWC module resolution algorithm doesn't match, tools might decide to fallback to the standard Node resolution.